include/IECore/IndexedIOAlgo.h

```diff
@@ -0,0 +1,92 @@
+//  met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Image Engine Design nor the names of any
+//       other contributors to this software may be used to endorse or
+//       promote products derived from this software without specific prior
+//       written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECore/Export.h"
+
+#include "IECore/IndexedIO.h"
+
+namespace IECore
+{
+	namespace IndexedIOAlgo
+	{
+		template<typename T>
+		struct FileStats
+		{
+			static constexpr unsigned int numBits = 64;
+			std::array<T, numBits> numBytes;
+			std::array<T, numBits> numBlocks;
+
+			FileStats()
+			{
+				for (unsigned int i = 0; i < numBits;++i)
+				{
+					numBytes[i] = 0UL;
+					numBlocks[i] = 0UL;
+				}
+			}
+
+			template<typename S>
+			FileStats(const FileStats<S> &other )
+			{
+				for (unsigned int i = 0; i < numBits;++i)
+				{
+					numBytes[i] = other.numBytes[i];
+					numBlocks[i] = other.numBlocks[i];
+				}
+
+			}
+
+			void addBlock(size_t blockSize)
+			{
+				for (unsigned int i = 0; i < numBits; ++i)
+				{
+					if ( blockSize <= (1U << i) )
+					{
+						numBlocks[i]++;
+						numBytes[i]+= blockSize;
+						break;
+					}
+				}
+			}
+		};
+
+		/// Recursively copy from 'src' to 'dst' using TBB if 'numThreads' > 1 or
+		/// using a regular recursive call otherwise.
+		/// Returns a histogram of blockSize and byte totals for 64 power 2 bins in FileStats
+		/// If 'dst' is null then only reads are performed on the 'src' IndexedIO and FileStats generated.
+		/// Note / todo : Currently IndexedIO doesn't support multithreaded writes so a copy will always use
+		/// the single threaded version.
+		IECORE_API FileStats<size_t> copy(ConstIndexedIOPtr src, IndexedIOPtr dst, int numThreads);
```

Comment:
This feels too overloaded to me :

- If I just want to do a copy, I most likely don't care about the stats, or want the overhead of collection. I also don't want to have to pass a numThreads argument that will be ignored.
- If I just want the stats, I'm unlikely to be looking around for a method called `copy()`. I most likely also want them as fast as possible, and I don't want to have to pass extra `nullPtr, numThreads` arguments to get them. The "TBB way" is to not pass thread counts into methods anyway, but to specify them externally when setting up the thread pool.
- If I'm just trying to do performance testing for reading, again, I'm not looking for a method called `copy()`, and it's not that intuitive to have to pass a `nullptr` for `dst`.

What do you think to something along these lines? It looks like you have everything structured nicely internally so that it might be quite easy to wrap it with this interface instead?

```
FileStats stats( const IndexedIO *io );
void copy( const IndexedIO *src, IndexedIO *dst );
void parallelReadAll( const IndexedIO *src );
```

Note that I've also used raw pointers for the arguments, since none of the methods need to take ownership.

Reply:
I couldn't find a python binding for task_scheduler_init. So how about an boost::optional thread count parameter which is used for performance testing but in other cases it's omitted?

Also, I'm not sure how useful blocks stats are beyond testing the file format so how about the following?

```
void copy( const IndexedIO *src, IndexedIO *dst );
FileStats  parallelReadAll( const IndexedIO *src, boost::optional<int> threadCount );
```

Reply:
> I couldn't find a python binding for task_scheduler_init.

There's one here :

https://github.com/GafferHQ/gaffer/blob/master/src/GafferModule/GafferModule.cpp#L101

We could move that to IECore if you liked, or you could just do your testing with `gaffer python -threads N testScript.py`, which is what I tend to do. Let's stick with the standard TBB idioms via one of these methods rather than monkey around with optional thread counts.

> Also, I'm not sure how useful blocks stats are beyond testing the file format so how about the following?

Sounds good.

Reply:
> We could move that to IECore if you liked, or you could just do your testing with gaffer python -threads N testScript.py, which is what I tend to do. Let's stick with the standard TBB idioms via one of these methods rather than monkey around with optional thread counts.

I don't think it's correct to require gaffer to test an IECore function and will take the option of moving the Gaffer wrapper to Cortex.